File: PythonVersion/source/MyQueue.py
from typing import *

from DataStructure.PythonVersion.source.MyConst import MAX_SIZE

T = TypeVar('T')

class MyQueue(Generic[T]):
    def __init__(self):
        self.__elements = [None for _ in range(MAX_SIZE)]
        self.__size = 0
    def __len__(self):
        return self.__size
    def is_empty(self) -> bool:
        return self.__size == 0
    def enqueue(self, element: T) -> None:
        self.__size += 1
        if self.__size > MAX_SIZE:
            self.__size -= 1
            raise OverflowError
        self.__elements[self.__size - 1] = element
    def dequeue(self) -> T:
        if self.__size < 1:
            raise IndexError('Queue is empty')
        e = self.__elements[0]
        for i in range(1, self.__size):
            self.__elements[i - 1] = self.__elements[i]
        self.__size -= 1
        return e
    def peek(self) -> T:
        if self.is_empty():
            raise IndexError('Queue is empty')
        return self.__elements[0]
    def __str__(self):
        sb = "["
        for i in range(self.__size):
            if i == self.__size - 1:
                sb += str(self.__elements[i])
            else:
                sb += str(self.__elements[i]) + ", "
        sb += "]"
        return sb

File: JavaVersion/source/MyQueue.java
package javaVersion.source;

import static javaVersion.source.MyConst.MAX_SIZE;

/**
 * 一个挺好用的队列~
 * @param <E>
 */
public class MyQueue<E>{
    private final E[] element;
    private int size;
    /**
     * 构造函数，构造一个初始的队列
     */
    public MyQueue() {
        element = (E[]) new Object[MAX_SIZE];
        for (int i = 0; i < element.length; i++)
            element[i] = null;
        size = 0;
    }
    /**
     * 获取长度
     * @return 长度
     */
    public int size() {
        return size;
    }
    /**
     * 判空
     * @return 是否为空
     */
    public boolean isEmpty() {
        return size == 0;
    }
    /**
     * 入队
     * @param value 值
     */
    public void enQueue(E value) {
        element[size] = value;
        size++;
    }
    /**
     * 出队
     * @return 值
     */
    public E deQueue() {
        if(size < 1) throw new ArrayIndexOutOfBoundsException("empty queue");
        E e = element[0];
        System.arraycopy(element, 1, element, 0, size);
        size--;
        return e;
    }
    /**
     * 检查队首值
     * @return 值
     */
    public E peek() {
        if(size <= 0) throw new ArrayIndexOutOfBoundsException("empty queue!");
        return element[0];
    }
    /**
     * toString函数
     * @return 格式化输出
     */
    @Override
    public String toString() {
        StringBuilder s = new StringBuilder("[");
        for (int i = 0; i < size; i++){
            if(i == size - 1){
                s.append(element[i]);
            }else{
                s.append(element[i]).append(", ");
            }
        }
        s.append("]");
        return s.toString();
    }
}
